### DISCUSSION TITLE
How to set minimum log levels?

### DISCUSSION THREAD

**Discussion by imbolc** [jmesmon/rust-systemd#176] [Q&A] [answered]
Is there a way to set different log levels for different crates (to filter verbose debug messages of some libs), something like `RUST_LOG` variable for `env_logger`? I can't find any configuration options after `systemd::journal::JournalLog::init()`.

**Comment by codyps** [comment]
The current journal `log` support in this crate is very minimal: it redirects all log messages to the journal without any filtering, and assumes one would use filtering when retrieving the logs. This is primarily because I didn't want to try to reproduce filtering here. And yes, this is likely way too noisy if some libraries that make heavy use of debug/trace messages are built in.

There may be other crates that allow control over the output to the systemd journal, or if not, one could write one that uses the APIs exposed in this crate to log output.

I've, in the past, thought about splitting out the support for `log` into a separate crate so that we'd be more free to add additional functionality to it.

Right now this is the extent of our logger:

https://github.com/jmesmon/rust-systemd/blob/8137b283bda83a04f58fbec7c936ca365c65a6cd/src/journal.rs#L85-L106

Modifying the `enabled()` entry would be needed to add hooks for filtering (whether from `env_logger` or elsewhere).

I'm open to accepting changes to the included logger that add some filtering functionality. Even better if there's some existing log filtering crate we can just pull in and call in `enabled()`.

**Comment by imbolc** [comment]
You're right, it makes more sense to add journal as a target to `env_logger` rather than copying all the filtering machinery here. I'll ask there, thank you.

**Comment by codyps** [comment]
I'm not sure env_logger really has a setup for plugable outputs: it appears to just expose stderr and stdout. There are other logging systems ([slog](https://crates.io/crates/slog), [tracing](https://crates.io/crates/tracing)) that do have plugable outputs and have some support for filtering in similar ways to env_logger. For those, I believe the approach is to create a new crate that adapts the logging framework (slog/tracing) to the output sink (systemd journal, in this case).

(Even if the solution you go with doesn't involve updating this systemd crate to add filtering, I'd be very interested in what you end up using)

**Comment by imbolc** [comment]
Yep, there's no plugins in env_logger, though they're already [discussing](https://github.com/env-logger-rs/env_logger/issues/178) adding a file [Target](https://docs.rs/env_logger/0.8.3/env_logger/enum.Target.html), so I've opened a similar [issue](https://github.com/env-logger-rs/env_logger/issues/194) on journald. Tracing already has [journald subscriber](https://github.com/tokio-rs/tracing/tree/master/tracing-journald). It just emotionally hard for me to bring tracing into a simple cli tool with a synchronous workflow :)

**Comment by imbolc** [comment, marked as answer]
Hey :) I ended up using tracing, here's the setup: https://gist.github.com/imbolc/f75f924f66c129f0e1e910f59fb75c0d

**Reply by codyps** [reply]
Cool, that looks like a good solution. The other nice benefit of `tracing-journald` is that it's a pure rust implimentation (ie: no need to link in libsystemd)
